---
title: "Microbial Signatures Data Munging"
package: MungeCuratedMGS
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: yes
    toc_depth: 4
abstract: |
  Just data munging.
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE)
```

# Installation

```{r eval=FALSE}
BiocManager::install('seandavi/MungeCuratedMGS')
```

# Data entities

```{r message=FALSE, warning=FALSE}
library(MungeCuratedMGS)
```

# Authentication

Run only once to authenticate, before running vignette (not evaluated in vignette):
```{r eval=FALSE}
library(googlesheets)
token = gs_auth()
saveRDS(token, file = "/tmp/googlesheets_token.rds")
```

Then run with the vignette, each time:
```{r eval=FALSE}
library(googlesheets)
suppressMessages(gs_auth(token = "/tmp/googlesheets_token.rds", verbose = FALSE))
```

## Ontologies

### EFO

```{r efo, cache=TRUE, eval=FALSE}
efo = get_efo()
head(efo)
```

### UBERON

```{r uberon, cache=TRUE, eval=FALSE}
uberon = get_uberon()
head(uberon)
```

Ontologies can be combined like so:

```{r, eval=FALSE}
library(dplyr)
full_ontologies = dplyr::bind_rows(uberon, efo)
```

## NCBI Taxonomy

### Names and Taxon IDs

```{r taxon_names, cache=TRUE, eval=FALSE}
taxon_dir = get_taxon_dump_files()
taxon_id_names = get_taxon_names(taxon_dir)
head(taxon_id_names, 10)
```

### obtain the set cMD features that map to an NCBI TaxID

```{r, eval = FALSE}
library(curatedMetagenomicData)
ds <- curatedMetagenomicData("*metaphlan*", dryrun = TRUE)
ds <- grep("WenC", ds, value = TRUE, invert = TRUE)
esets <- curatedMetagenomicData(ds, dryrun = FALSE)
feats <- lapply(esets, rownames)
allfeats <- Reduce(union, feats)
cat(allfeats, file = "../inst/extdata/cMD_allfeats.txt", sep = "\n")
```

```{r}
feat.file <- system.file("extdata/cMD_allfeats.txt", package = "MungeCuratedMGS")
allfeats <- scan(feat.file, what = "character")
m2n <- metaphlan2ncbi(metaphlan.version = "2.0")
m2n.cmd <- m2n[allfeats] 
m2n.cmd <- m2n.cmd[!is.na(m2n.cmd)]
```

```{r, eval = FALSE}
dat <- cbind(names(m2n.cmd), unname(m2n.cmd))
colnames(dat) <- c("metaphlan", "ncbi")
write.csv(dat, file = "../inst/extdata/cMD_metaphlan2ncbi.csv", 
            quote = FALSE, row.names = FALSE)
```

## Data export

### Write all curation tables as csv files

```{r}
sheet <- curation_sheet()
dim(sheet)
```

```{r}
study <- create_study_table(sheet)
experiment <- create_experiment_table(sheet)
signature <- create_signature_table(sheet)
```

Create keys and write to file:
```{r}
freeform.cols <- c("antibiotics exclusion", "Group 1 definition", "source", "description")

sig.keys <- paste("Signature", seq_len(nrow(signature)))
exp.cols <- setdiff(colnames(experiment), freeform.cols)
exp.keys <- create_keys("Experiment ", experiment[,exp.cols])
stud.cols <- setdiff(colnames(study), freeform.cols)
stud.keys <- create_keys("Study ", study[,stud.cols])

# reset signature counter (in signature table)
sig.rle <- rle(exp.keys)$lengths
sig.rle <- lapply(sig.rle, seq_len)
sig.keys <- paste("Signature", unlist(sig.rle))

# reset experiment counter (in signature table)
exp.spl <- split(exp.keys, as.integer(sub("Study ", "", stud.keys)))
exp.rle <- lapply(exp.spl, function(e) rle(e)$lengths)
exp.sig.keys <- lapply(exp.rle, function(e) rep(seq_along(e), e))
exp.sig.keys <- unlist(exp.sig.keys)

signature$Experiment <- paste("Experiment", exp.sig.keys)
signature$Study <- stud.keys
experiment$Study <- stud.keys

signature$`Page Name` <- sig.keys
experiment$`Page Name` <- signature$Experiment
study$`Page Name`<- stud.keys 

experiment <- experiment[!duplicated(experiment[,exp.cols]),]
study <- unique(study)

# reset experiment counter (in experiment table)
# exp.rle <- rle(experiment$Study)$lengths
# exp.rle <- lapply(exp.rle, seq_len)
# experiment[["Page Name"]] <- paste("Experiment", unlist(exp.rle))

write.csv(study, file="studies.csv", row.names = FALSE)
write.csv(experiment, file="experiments.csv", row.names = FALSE)
write.csv(signature, file="signatures.csv", row.names = FALSE)
```

Double-check via joins:
```{r}
names(experiment)[28] <- "Experiment"
names(study)[6] <- "Study"
sig.exp <- plyr::join(experiment, signature, by = c("Study", "Experiment"))
ses <- plyr::join(study,sig.exp, by = "Study")
dupl <- study$PMID[duplicated(study$PMID)]
dupl.studs <- study[study$PMID %in% dupl,]
dupl.studs[order(dupl.studs$PMID),]
```

Write the signatures file again, alternating metaphlan and NCBI columns:
```{r}
nonsigcols <- grep("Metaphlan|NCBI|Study|Page", colnames(signature), invert=TRUE)
sigcols <- as.integer(t(cbind(grep("Metaphlan", colnames(signature)), grep("NCBI", colnames(signature)))))
keys <- grep("Study|Page", colnames(signature))
signature2 <- signature[, c(nonsigcols, sigcols, keys)]
write.csv(signature2, file="signatures_alternatingNCBI.csv", row.names = FALSE)
```

Single PMIDs split into multiple studies
```{r}
pmidsplits <- table(study$PMID)
pmidsplits <- pmidsplits[pmidsplits > 1]
pmidsplits <- names(pmidsplits)
splits <- study[study$PMID %in% pmidsplits, ]
(splits <- splits[order(splits$PMID), ])
write.csv(splits, "splitstudies.csv", row.names = FALSE)
```

```{r, eval=FALSE}
signature_ike <- readr::read_csv("~/Downloads/Signature_ike.csv")
study_ike <- readr::read_csv("~/Downloads/Studies_ike.csv")
```

```{r, eval=FALSE}
colnames(signature_ike)[!colnames(signature_ike) %in% colnames(signature)] %>%
  grep(pattern="metaphlan|ncbi", x=., invert = TRUE, ignore.case = TRUE, value = TRUE)
```
```{r, eval=FALSE}
colnames(signature)[!colnames(signature) %in% colnames(signature_ike)] %>%
  grep(pattern="metaphlan|ncbi", x=., invert = TRUE, ignore.case = TRUE, value = TRUE)
```

```{r, eval=FALSE}
colnames(study_ike)[!colnames(study_ike) %in% colnames(study)] %>%
  grep(pattern="metaphlan|ncbi", x=., invert = TRUE, ignore.case = TRUE, value = TRUE)
```

```{r, eval=FALSE}
colnames(study)[!colnames(study) %in% colnames(study_ike)] %>%
  grep(pattern="metaphlan|ncbi", x=., invert = TRUE, ignore.case = TRUE, value = TRUE)
```

